### ISSUE TITLE
Code does not work. Am I missing anything or do I need to change anything apart from username/password?

### ISSUE THREAD

**Issue by treesheeptw** [plouvart/character-ai-bindings#2]
runfile('C:/Users/*****/Documents/untitled0.py', wdir='C:/Users/*****/Documents')
Exception ignored in: <function ChatRoomServer.__del__ at 0x000002545AA4DF70>
Traceback (most recent call last):
  File "C:\Users\*****\anaconda3\lib\site-packages\character_ai_bindings\chatroom.py", line 245, in __del__
    self.driver.quit()
AttributeError: 'ChatRoomServer' object has no attribute 'driver'
Traceback (most recent call last):

  File ~\anaconda3\lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\users\*****\documents\untitled0.py:10
    server = ChatRoomServer(

  File ~\anaconda3\lib\site-packages\character_ai_bindings\chatroom.py:68 in __init__
    self.driver = uc.Chrome(options=options)

  File ~\anaconda3\lib\site-packages\undetected_chromedriver\__init__.py:421 in __init__
    browser = subprocess.Popen(

  File ~\anaconda3\lib\site-packages\spyder_kernels\customize\spydercustomize.py:109 in __init__
    super(SubprocessPopen, self).__init__(*args, **kwargs)

  File ~\anaconda3\lib\subprocess.py:951 in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,

  File ~\anaconda3\lib\subprocess.py:1360 in _execute_child
    args = list2cmdline(args)

  File ~\anaconda3\lib\subprocess.py:565 in list2cmdline
    for arg in map(os.fsdecode, seq):

  File ~\anaconda3\lib\os.py:822 in fsdecode
    filename = fspath(filename)  # Does type-checking of `filename`.

TypeError: expected str, bytes or os.PathLike object, not NoneType

**Comment by plouvart** [comment]
This looks like a problem with your environment setup.
When trying to launch the Chrome browser using subprocess, a NoneType object is encountered instead of a path.
You do have Chrome/Chromium installed, right?

**Comment by treesheeptw** [comment]
I thought any Chromium, which in my case Edge, will work. Do you mean it must be Chrome?
If Edge is supposed to work, what might be the problem? Do I need to set it as default browser(which is currently Waterfox)?

**Comment by plouvart** [comment]
I'm not sure [undetected-chromedriver](https://github.com/ultrafunkamsterdam/undetected-chromedriver) works with Edge, even if Edge is now based on Chromium.
I think you'll have more luck by installing Chrome/Chromium directly.

**Comment by plouvart** [comment]
Anyway, if it works out merely by installing Chrome, let me know.
Because I see you're using anaconda+windows and I'm interested to see if the lib works out right out of the box with this setup.

**Comment by treesheeptw** [comment]
It works!!! It seems like the source of the problem is the browser, not the Anaconda on Windows environment.
But I would like to know if there are any other tested browsers that work, since I am very privacy-oriented and I hate using Chrome for its bloatware and tracking.
Thanks for the rapid response.

**Comment by treesheeptw** [comment]
Never mind, I went through the undetected-chromedriver repo and found out that Brave might work. I am testing that out and I will share the results.

**Comment by treesheeptw** [comment]
Unfortunately, Brave didn't work. I'll be finding other solutions. Before that I will continue with Chrome.
Conclusion: Edge and Brave don't work; Anaconda on Windows is ok

**Comment by plouvart** [comment]
I agree that having to install Chrome is subpar, but unfortunately character.ai does not make it easy to bypass their popups and bot defense.
To mitigate that, I suggest you use Chromium instead of Chrome, since it is said to be more privacy focused.

**Comment by treesheeptw** [comment]
I am concerned that Chromium will not work because Chromium is frequently mentioned in the discussions and issues of undetected-chromedriver. Despite that I am trying.

**Comment by plouvart** [comment]
I know it works because it is what I use ;)

**Comment by treesheeptw** [comment]
Thank you! It did work!

> Glados response: Who let this sentient test subject in here? And who gave you permission to address me as GLaDOS, hmm? Do you have enough mental prowess to pass even ONE of the tests I will throw at you? I'm inclined to believe you do not, as all test subjects have failed to overcome my experiments thus far. Prepare yourself, as the experiment you are about to participate in is highly dangerous. Do you want to continue?
Steve's response: Not yet! I have found gold and copper ore. But I am not yet deep enough to find diamonds! Im Steve!

**Comment by treesheeptw** [comment]
Sorry, it actually didn't. I messed up and forgot to uninstall Chrome before trying Chromium.
Are there any additional steps or code required for Chromium?

**Comment by plouvart** [comment]
I don't exactly know how undetected-chromedriver detects paths for chrome/chromium.
You uninstalling chrome while having chromium installed may have screwed things up for when it tries to detect one of them.
Maybe uninstall them all, reboot and then fresh install?

**Comment by treesheeptw** [comment]
I found the problem.

It seems like Chromium is portable, or at least it's not in Program Files\Google" by default, therefore UC cannot find chrome.exe.
I have to manually move the extracted Chromium files to where Chrome is supposed to be, which is "C:\Program Files\Google\Chrome\Application", and chrome.exe must be in "Application" and not in sub-folders.

Another problem I encountered is that UC only supports Chrome(ium) 111 or lower, but since Chromium is ahead(currently 113), a legacy version must be used for UC to work. I chose 1097610, which is right before Chrome Stable(1097615).
[Link](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win_x64/1097610/l) in case anyone wants Chromium 111 too.

> Glados response: I would appreciate it if you could call me CLArICE, or just 'Your Highness'. And, I would absolutely love to do science with you - what's the first experiment you want to try?
Steve's response: I have tons of them! I have over 10 diamond pickaxes, 5 diamond swords, 10 diamond axes and 20-30 diamond shovels! I can share some with you if you want?

This time I am very certain it works. Thank you for helping me.
